### ISSUE TITLE
Extending functionality of EnsembleModel / integration with SampleFitCombine.jl

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/MLJ.jl#363]
I'm starting this thread to brainstorm the possibility of extending the functionality of `EnsembleModel` beyond bagging, and in particular to discuss the proposal below to integrate the functionality already provided by [SampleFitCombine.jl](https://github.com/JockLawrie/SampleFitCombine.jl). The author of this package @JockLawrie has offered his assistance. Here is his proposal, copied from [this thread](https://github.com/alan-turing-institute/MLJ.jl/issues/317#issuecomment-556677564):

@JockLawrie says:

Yes the homogenous ensembles provide some of the functionality. I think it could be more general, as per `SampleFitCombine`, which also allows other ways to sample training data and combine learners.

For example, see the test case which partitions the training data by clustering the predicted training data according to `cdf(predict(ensemble, training_data), training_data_point)`. This is the Sample step. Then a learner is fitted to each cluster of data points in the partition (Fit), then the learners are combine with weights that minimize the loss (Combine).

I'm happy to try and make contributions, but at this point I think it better to flesh out what it would look like first. `SampleFitCombine` is my proposal, namely:
1. The `Ensemble` struct
2. Sample methods constructed independently, using `wsample!` in `StatsBase` for example.
3. Fit methods are already provided by MLJ.
4. Three `combine` methods (uniform, pre-specified, optimal)
5. `predict(ensemble, Xtest)`
6. `loss(ensemble, X, y, lossfunction)`
7. Conveniences

The `Ensemble` struct in `SampleFitCombine` is serves the same purpose as MLJ's `EnsembleModel`. However the latter seems to assume that bagging is the primary/only mechanism for constructing ensembles. But training data may be sampled in many ways, not just via bagging. And learners may be fit independently or in sequence. E.g., boosting can be expressed as a sequential sample-fit-combine method.

**Comment by ablaom** [comment]
@JockLawrie Thanks for the offer of help. Your proposal sounds promising. I think it would be good to have a call to get things rolling. Will contact you on slack.

**Comment by azev77** [comment]
To clarify what I meant in #466. 
My understanding is there are 3 broad categories of ensembles: bagging/boosting/stacking. 

`SuperLearner` allows you to train a large number of models, w/ possibly multiple grids per model & then reports the out-of-sample scores for each model & different stacked ensembles of the underlying models: 
![image](https://user-images.githubusercontent.com/7883904/76692320-26467500-662b-11ea-9eb2-5c829eef49b7.png)

I have a program that computes all 48 MLJ regressions that match(X, y). I'd like to automatically compare the scores of different stacked ensembles of these models. 
That is I'd like to automate what you're doing in [stacking](https://alan-turing-institute.github.io/MLJTutorials/getting-started/stacking/) for a large number of models.
This isn't a priority right now, just wanted you guys to be aware of `SuperLearner`.

**Comment by azev77** [comment]
Also, suppose I train 3 models in MLJ {m1, m2, m3}, then for each model I compute the out of sample residuals {e1, e2, e3} is there a convenient way to create the plot above for these residuals w/ the model names on the y-axis? 
Also w/ a `dot` @ the RMSE/MAE/MAPE etc

**Comment by azev77** [comment]
Also might be worth checking out the different ensemble options in @ppalmes in https://github.com/ppalmes/TSMLextra.jl.
